### PR TITLE
[DF] Improve progress bar finalization

### PR DESCRIPTION
**Pull request by martamaja10**
After a quick discussion at one of the recent PPP round tables, we decided that the ProgressBar feature is missing a nice finalization, which prints the final status with all processed events. This PR introduces this feature. It now looks as on the attached screen-shot (both for MT and non MT modes, checked with df106_HiggsToFourLeptons and df102_NanoAODDimuonAnalysis tutorials). 
Screenshot 2023-10-13 at 15 03 30

### CODE REVIEW

**Comment by pcanal** [comment]
In the picture I see 2 stats lines, shouldn't we keep only one?

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx [hunk 1]
```diff
@@ -298,6 +298,8 @@ private:
    double EvtPerSec() const;
    std::pair<std::size_t, std::chrono::seconds> RecordEvtCountAndTime();
    void PrintStats(std::ostream &stream, std::size_t currentEventCount, std::chrono::seconds totalElapsedSeconds) const;
+   void PrintStats_final(std::ostream &stream, std::size_t currentEventCount,
```
Comment (vepadulano) [inline tree/dataframe/inc/ROOT/RDFHelpers.hxx:301]:
```suggestion
   void PrintStatsFinal(std::ostream &stream, std::size_t currentEventCount,
```

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx [hunk 2]
```diff
@@ -381,13 +385,11 @@ public:
...
       // We only print every n seconds.
       if (duration_cast<seconds>(system_clock::now() - fLastPrintTime) < fPrintInterval) {
-
          // Unless we are at the end of file processing, then we want to print the progress bar again (the final status)
          // Otherwise, if the last processed files are too small and they are processed in less than the time interval,
          // the final progress bar status would be incomplete. We want to prevent this from happening.
-
          if (fTotalFiles != currentFileIdx) {
-            if (currentFileIdx <= GetNEventsOfCurrentFile - fIncrement) {
+            if (GetNEventsOfCurrentFile - fProcessedEvents > fIncrement) {
```
Comment (vepadulano) [inline tree/dataframe/inc/ROOT/RDFHelpers.hxx:392]: Can you comment on why we need the change in this logic here? Maybe also add the comment inline.

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx [hunk 3]
```diff
@@ -431,7 +433,7 @@ public:
       return fSampleNameToEventEntries.size();
    }
 };
-
+} // namespace Experimental
```
Comment (vepadulano) [inline tree/dataframe/inc/ROOT/RDFHelpers.hxx:436]: Could you put the changing of namespace of the `ProgressHelper` class in a separate commit? And thanks for noticing it, this slipped in my reviews of the previous PR

File: tree/dataframe/src/RDFHelpers.cxx [hunk 4]
```diff
@@ -258,14 +260,47 @@ void ProgressHelper::PrintStats(std::ostream &stream, std::size_t currentEventCo
       std::chrono::seconds remainingSeconds(
          static_cast<long long>((ComputeNEventsSoFar() - currentEventCount) / evtpersec));
       stream << " " << remainingSeconds << " "
-             << " remaining time (per file)";
+             << " remaining time (per file being processed)";
```
Comment (vepadulano) [inline tree/dataframe/src/RDFHelpers.cxx:263]: Maybe also this change should go in a separate commit for consistency

File: tree/dataframe/src/RDFHelpers.cxx [hunk 5]
```diff
@@ -258,14 +260,47 @@ void ProgressHelper::PrintStats(std::ostream &stream, std::size_t currentEventCo
...
    }
 
    stream << "]   ";
 }
 
+void ProgressHelper::PrintStats_final(std::ostream &stream, std::size_t currentEventCount,
+                                      std::chrono::seconds elapsedSeconds) const
+{
+   auto evtpersec = EvtPerSec();
+   auto GetNEventsOfCurrentFile = ComputeNEventsSoFar();
```
Comment (vepadulano) [inline tree/dataframe/src/RDFHelpers.cxx:275]:
```suggestion
   auto totalEvents = ComputeNEventsSoFar();
```

File: tree/dataframe/src/RDFHelpers.cxx [hunk 6]
```diff
@@ -258,14 +260,47 @@ void ProgressHelper::PrintStats(std::ostream &stream, std::size_t currentEventCo
...
+   auto totalFiles = fTotalFiles;
+
+   std::chrono::seconds remainingSeconds(
+      static_cast<long long>((ComputeNEventsSoFar() - currentEventCount) / evtpersec));
```
Comment (vepadulano) [inline tree/dataframe/src/RDFHelpers.cxx:279]: Here we should be able to reuse the variable just taken above right?
```suggestion
      static_cast<long long>((totalEvents - currentEventCount) / evtpersec));
```

File: tree/dataframe/src/RDFHelpers.cxx [hunk 7]
```diff
@@ -308,7 +342,23 @@ class ProgressBarAction final : public ROOT::Detail::RDF::RActionImpl<ProgressBa
...
+         return;
+      std::lock_guard<std::mutex> lockGuard(fPrintMutex, std::adopt_lock);
+      std::size_t eventCount;
+      std::chrono::seconds elapsedSeconds;
+      std::tie(eventCount, elapsedSeconds) = fHelper->RecordEvtCountAndTime();
+
+      fHelper->PrintProgressbar(std::cout, eventCount);
+      fHelper->PrintStats_final(std::cout, eventCount, elapsedSeconds);
+
+      std::cout << '\n';
```
Comment (vepadulano) [inline tree/dataframe/src/RDFHelpers.cxx:360]: I would rewrite a bit this function with the following ideas:
* Remove the extra `\n` at the beginning, so we can try to overwrite the same line were the progress bar was previously written
* Avoid running `PrintProgressbar` here, since it's anyway already at the end, we are interested in the final stats mostly
* Use some magic `std::cout << '\r' << std::string(get_tty_size(), ' ') << '\r';` to reset the current output line before writing the final stats. I tested this on my machine, could you also test it on Mac?

```suggestion
      std::mutex fPrintMutex;
      if (!fPrintMutex.try_lock())
         return;
      std::lock_guard<std::mutex> lockGuard(fPrintMutex, std::adopt_lock);
      const auto &[eventCount, elapsedSeconds] = fHelper->RecordEvtCountAndTime();

      // The next line resets the current line output in the terminal.
      // Brings the cursor at the beginning ('\r'), prints whitespace with the
      // same length as the terminal size, then resets the cursor again so we
      // can print on a clean line the final stats.
      std::cout << '\r' << std::string(get_tty_size(), ' ') << '\r';
      fHelper->PrintStats_final(std::cout, eventCount, elapsedSeconds);

      std::cout << '\n';
```

**Review by vepadulano — CHANGES_REQUESTED** [review]
Thanks @martamaja10 , I believe the changes go in the right direction! See some comments inline. Also, something which came up while I was reviewing, I believe we shoud rename the function `AddProgressbar` to `AddProgressBar`, since "progress bar" is written as two separate words and we capitalize every starting letter in our convention.

File: tree/dataframe/inc/ROOT/RDFHelpers.hxx (hunk 2 above)
Comment (martamaja10) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:392]: the logic is kind of described in the comment above (but maybe I can improve it, it's true): if you print every second, at some point you process events too fast and you end up not really printing anymore so when you introduce the second statement, this is not true anymore. The change is due a bug in the implementation - I was comparing file index to the total number of events in a file instead of comparing the number of processed events to the total number of events (as it is done now). I also changed the format of the equation so it matches the one above for printing each second. Now, actually if we decide to anyways replace the final progress bar with the end state statement (and only print a one-liner), maybe this is not needed at all, let me test things and see what it looks like.

Comment (martamaja10) [reply tree/dataframe/inc/ROOT/RDFHelpers.hxx:392]: actually now with the addition of the final status line, this trick is not needed anymore, I'll remove it

File: tree/dataframe/src/RDFHelpers.cxx [hunk 8]
```diff
@@ -258,14 +257,46 @@ void ProgressHelper::PrintStats(std::ostream &stream, std::size_t currentEventCo
...
 }
 
+void ProgressHelper::PrintStatsFinal(std::ostream &stream, std::size_t currentEventCount,
+                                     std::chrono::seconds elapsedSeconds) const
+{
+   auto evtpersec = EvtPerSec();
+   auto totalEvents = ComputeNEventsSoFar();
+   auto totalFiles = fTotalFiles;
+
+   std::chrono::seconds remainingSeconds(static_cast<long long>((totalEvents - currentEventCount) / evtpersec));
```
Comment (vepadulano) [inline tree/dataframe/src/RDFHelpers.cxx:275]: It looks like this variable is not used in the function
```suggestion
```

**Review by vepadulano — APPROVED** [review]
Thanks, looks great! Just a small comment for cleanup

File: tree/dataframe/src/RDFHelpers.cxx (hunk 8 above)
Comment (martamaja10) [reply tree/dataframe/src/RDFHelpers.cxx:275]: true, thanks!
